packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2651,9 +2665,9 @@ export class KeyringController extends BaseController<
         return await callback({ releaseLock });
       } catch (e) {
         // Keyrings and password are restored to their previous state
-        await this.#restoreSerializedKeyrings(currentSerializedKeyrings);
-        this.#password = currentPassword;
         this.#keyringsMetadata = currentKeyringsMetadata;
+        this.#password = currentPassword;
+        await this.#restoreSerializedKeyrings(currentSerializedKeyrings);
```

Comment:
we must restore metadata first, or `#restoreSerializedKeyrings` will throw an error because it could find new metadata added by the failed operation